runtime/major_gc.c

```diff
@@ -1331,6 +1355,92 @@ static intnat ephe_sweep (caml_domain_state* domain_state, intnat budget)
   return budget;
 }
 
+static void request_mark_phase (void)
+{
+  if (caml_gc_phase == Phase_sweep_main &&
+      atomic_load_relaxed(&caml_gc_mark_phase_requested) == 0)
+    atomic_store_release(&caml_gc_mark_phase_requested, 1);
+}
+
+void caml_mark_roots_stw (int participant_count,
+                          caml_domain_state** barrier_participants)
+{
+  if (caml_gc_phase != Phase_sweep_main)
+    return;
+
+  enum global_roots_status {
+    WORK_UNSTARTED,
+    WORK_STARTED,
+    WORK_COMPLETE
+  };
+  static atomic_uintnat global_roots_started;
+
+  Caml_global_barrier_if_final(participant_count) {
+    caml_gc_phase = Phase_sweep_and_mark_main;
+    atomic_store_relaxed(&global_roots_started, WORK_UNSTARTED);
+  }
+
+  caml_domain_state* domain = Caml_state;
+
+  /* Ephemerons */
+#ifdef DEBUG
+  orph_ephe_list_verify_status (caml_global_heap_state.UNMARKED);
+#endif
+  /* Adopt orphaned work from domains that were spawned and terminated in the
+     previous cycle. */
+  adopt_orphaned_work (caml_global_heap_state.UNMARKED);
+  begin_ephe_marking();
+
+  CAML_EV_BEGIN(EV_MAJOR_MARK_ROOTS);
+  {
+    uintnat work_unstarted = WORK_UNSTARTED;
+    if (atomic_load_relaxed(&global_roots_started) == WORK_UNSTARTED &&
+        atomic_compare_exchange_strong(&global_roots_started,
+                                       &work_unstarted, WORK_STARTED)) {
+      /* This domain did the CAS, so this domain marks the roots */
+      caml_scan_global_roots(&caml_darken, domain);
+      atomic_store_release(&global_roots_started, WORK_COMPLETE);
+    }
+  }
+  /* Locals, C locals, systhreads & finalisers */
+  caml_do_roots (&caml_darken, darken_scanning_flags, domain, domain, 0);
+  CAML_EV_END(EV_MAJOR_MARK_ROOTS);
+
+  CAML_EV_BEGIN(EV_MAJOR_MEMPROF_ROOTS);
+  caml_memprof_scan_roots(caml_darken, darken_scanning_flags, domain,
+                          domain, false);
+  CAML_EV_END(EV_MAJOR_MEMPROF_ROOTS);
+
+  caml_gc_log("Marking started, %ld entries on mark stack",
+              (long)domain->mark_stack->count);
+
+  if (domain->ephe_info->todo == (value) NULL)
+    ephe_todo_list_emptied();
+
+  /* Wait until global roots are marked. It's fine if other domains are still
```

Comment:
Can you explain why?

Reply:
I've added an "Ambitious TODO" comment here.